spinnaker_components/filter/filter.c

```diff
@@ -0,0 +1,239 @@
+/* Parallel implementation of the filter operator
+ * ----------------------------------------------
+ *
+ * Each core in the parallel filter is responsible for receiving packets for a
+ * subspace of the value represented by the filter and for transmitting
+ * transformed packets for a different subspace.  The general mode of operation
+ * on the timer update is:
+ *
+ *  1. Filter received values
+ *  2. Write filtered value into SDRAM
+ *  3. Read whole vector from SDRAM
+ *  4. Apply local portion of the transform and transmit transformed packets
+ *
+ * ---
+ *
+ * The following SDRAM regions are expected:
+ *
+ *  1. System region (see `filter_parameters_t`)
+ *  2. Output keys
+ *  3. Filter parameters
+ *  4. Filter routes
+ *  5. Transform
+ *
+ * ---
+ */
+
+/*****************************************************************************/
+#define MAX_VALS 1024  // 4kiB per page (in words)
+#define N_PAGES 8      // 8 pages
+#define PAGE_MASK 0x7  // Mask for the pages
+
+#include "../filter/filter_common.h"
+#include "input_filtering.h"
+#include "common-impl.h"
+
+// Global variables
+// General parameters (system region)
+typedef struct filter_parameters
+{
+  uint32_t machine_timestep;  // Machine time step / useconds
+
+  uint32_t whole_dimensions;  // Size of the entire vector
+
+  struct
+  {
+    uint32_t offset;  // Index of first dimension
+    uint32_t n_dims;  // Number of dimensions
+  } input_subspace;  // Subspace of the input managed by this instance
+
+  uint32_t output_subspace_n_dims;  // Size of output
+
+  value_t *vector;  // Shared input vector (pointer to SDRAM)
+} filter_parameters_t;
+filter_parameters_t parallel_filter;
+
+if_collection_t filters;  // Locally applied filters
+
+value_t *vector_subspace; // First item of our subspace of the vector (SDRAM)
+
+value_t *input;        // The complete input vector (copy in DTCM)
+
+pages_t pages;  // Pages of the transform
+value_t *transform;  // Our portion of the transform matrix (in SDRAM)
+
+uint32_t *keys;       // Multicast keys
+/*****************************************************************************/
+
+/*****************************************************************************/
+// Schedule writing out the filtered subspace of the whole vector that we
+// manage locally.
+void write_filtered_vector(void)
+{
+  // Copy our subspace of the input vector into the shared SDRAM representation
+  // of that vector.
+  spin1_dma_transfer(
+    WRITE_FILTERED_VECTOR << 30,  // Tag
+    vector_subspace,              // SDRAM address
+    filters.output,               // DTCM address
+    DMA_WRITE,                    // Direction
+    parallel_filter.input_subspace.n_dims * sizeof(value_t)  // Size
+  );
+}
+/*****************************************************************************/
+
+/*****************************************************************************/
```

Comment:
I can't see either of these functions being called from anywhere else so I think it would be clearer to move their bodies into ``dma_transfer_done``